libs/git/clone.go

```diff
@@ -0,0 +1,147 @@
+package git
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/databricks/cli/libs/cmdio"
+	"github.com/databricks/cli/libs/zip"
+)
+
+var errNotFound = errors.New("not found")
+
+type RepositoryNotFoundError struct {
+	url string
+}
+
+func (err RepositoryNotFoundError) Error() string {
+	return fmt.Sprintf("repository not found: %s", err.url)
+}
+
+func (err RepositoryNotFoundError) Is(other error) bool {
+	return other == errNotFound
+}
+
+type CloneOptions struct {
+	// Name of the organization or profile with the repository
+	Organization   string
+	RepositoryName string
+
+	// Git service provider. Eg: github, gitlab
+	Provider string
+
+	// Branch or tag name to clone
+	Reference string
+
+	// Path to clone into. The repository is cloned as ${RepositoryName}-${Reference}
+	// in this target directory.
+	TargetDir string
+}
+
+func (opts CloneOptions) repoUrl() string {
+	return fmt.Sprintf(`https://github.com/%s/%s`, opts.Organization, opts.RepositoryName)
```

Comment:
It assumes it's always github but it can be GitLab for example or any other provider, right?

Reply:
Yes, we can add support for gitlab as a followup. gitlab also support zip downloads, and for other providers like bitbucket we can defer to the CLI